### ISSUE TITLE
BC error with this PBES2 PKCS12 file

### ISSUE THREAD

**Issue by space88man** [bcgit/bc-java#1291]
When BC has higher priority than SUN there is an error when reading a PBES2 PKCS12 file:

```
# use BCPKCS12
java -cp ../lib/bcprov-jdk18on-172.jar -Djava.security.properties=../lib/java.security.test sun.security.tools.keytool.Main \
-list -keystore testks.p12 -storepass changeit -storetype BCPKCS12

Keystore type: BCPKCS12
Keystore provider: BC

Your keystore contains 1 entry

naverglobalrootcertificationauthority, 30 Nov 2022, trustedCertEntry, 
Certificate fingerprint (SHA-256): 88:F4:38:DC:F8:FF:D1:FA:8F:42:91:15:FF:E5:F8:2A:E1:E0:6E:0C:70:C3:75:FA:AD:71:7B:34:A4:9E:72:65
keytool error: java.io.IOException: keystore password was incorrect
```

```
# use PKCS12
java -cp ../lib/bcprov-jdk18on-172.jar -Djava.security.properties=../lib/java.security.test sun.security.tools.keytool.Main \
-list -keystore testks.p12 -storepass changeit

keytool error: java.io.IOException: keystore password was incorrect
```

```
# java.security.test
security.provider.1=sun.security.provider.Sun
security.provider.2=org.bouncycastle.jce.provider.BouncyCastleProvider
security.provider.3=sun.security.rsa.SunRsaSign
security.provider.4=sun.security.ec.SunEC
security.provider.5=com.sun.net.ssl.internal.ssl.Provider
security.provider.6=com.sun.crypto.provider.SunJCE
security.provider.7=sun.security.jgss.SunProvider
security.provider.8=com.sun.security.sasl.Provider
security.provider.9=org.jcp.xml.dsig.internal.dom.XMLDSigRI
security.provider.10=sun.security.smartcardio.SunPCSC
security.provider.11=sun.security.pkcs11.SunPKCS11 ${java.home}/lib/security/nss.cfg
```

[UPDATE] SunJCE / BC have different behaviour wrt PBEParameterSpec, and SunJSSE is hardcoded to expect SunJCE.

[testks.zip](https://github.com/bcgit/bc-java/files/10118364/testks.zip)

**Comment by space88man** [comment]
OpenSSL info 

```
$ openssl pkcs12 -in testks.p12 -passin pass:changeit  -info -noout
MAC: sha256, Iteration 10003
MAC length: 32, salt length: 20
PKCS7 Encrypted data: PBES2, PBKDF2, AES-256-CBC, Iteration 10002, PRF hmacWithSHA256
Certificate bag
```

**Comment by space88man** [comment]
I have narrowed the issue down to this in PBES2:

The SunJSSE PCKS12 Provider explicitly uses a string like "PBEWithHmacSHA256AndAES_256" to instantiate the Cipher.
Since this is not implemented by BC the provider is SunJCE... however the SecretKeyFactory of BC and Sun generate a different intermediate PBE key:

* BC has done the (add zero byte at the end) + UTF-16BE encoding of the passphrase
* SunJCE uses the password raw

```
// SunJCE / BC return different values here
private SecretKey getPBEKey(char[] password) throws IOException {
        SecretKey skey = null;

        try {
            PBEKeySpec keySpec = new PBEKeySpec(password);
            SecretKeyFactory skFac = SecretKeyFactory.getInstance("PBE");
            System.out.println(String.format("skFac provider = %s", skFac.getProvider()));
            skey = skFac.generateSecret(keySpec);
            keySpec.clearPassword();
        } catch (Exception e) {
            throw new IOException("getSecretKey failed: " + e.getMessage(), e);
        }
        return skey;
    }

```

**Comment by space88man** [comment]
TL;DR: don't mix SunJSSE with BC has SunJSSE has certain hardcoded behaviour expecting SunJCE.
